File: Top_100_Liked_Questions/105._Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
                                
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        # recursive postorder traversal
        preorder_idx = [0]
        
        def findNodeAndConnectChildren(inorder_left, inorder_right):
            if inorder_left >= inorder_right:
                return None
            
            preorder_val = preorder[preorder_idx[0]]
            preorder_idx[0] += 1
            
            i_idx = None
            for i in range(inorder_left, inorder_right):
                if inorder[i] == preorder_val:
                    i_idx = i
                    break
            
            if i_idx is None:
                raise Exception("{} not found in inorder {}".format(preorder_val, inorder[inorder_left:inorder_right]))
                
            left = findNodeAndConnectChildren(inorder_left, i_idx)
            right = findNodeAndConnectChildren(i_idx+1, inorder_right)
            return TreeNode(preorder_val, left, right)
        
        return findNodeAndConnectChildren(0, len(inorder))

File: Top_100_Liked_Questions/105._Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int[] preorder;
    public int[] inorder;
    public int preorder_idx;

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        this.preorder = preorder;
        this.inorder = inorder;
        this.preorder_idx = 0;

        return this.findNodeFromInorderAndConnect(0, inorder.length);
    }

    public TreeNode findNodeFromInorderAndConnect(int left, int right) {
        if (left >= right)
            return null;

        int inorder_idx = -1;
        int preorder_val = this.preorder[this.preorder_idx++];

        for (int i=left; i<right; i++) {
            if (this.inorder[i] == preorder_val) {
                inorder_idx = i;
                break;
            }
        }

        TreeNode node = new TreeNode(preorder_val);
        node.left = this.findNodeFromInorderAndConnect(left, inorder_idx);
        node.right = this.findNodeFromInorderAndConnect(inorder_idx+1, right);

        return node;
    }
}